txpipe/source_selector.py

```diff
@@ -619,14 +620,17 @@ def select_2d(self, data, is_2d=True):
         f2 = sel.sum() / n0
         sel &= s2n>s2n_cut
         f3 = sel.sum() / n0
+        sel &= data['zbin'] >= 0
+        f4 = sel.sum() / n0
 
         # Print out a message.  If we are selecting a 2D sample
         # this is the complete message.  Otherwise if we are about
         # to also apply a redshift bin cut about then the message will continue
         # as above
         if verbose and is_2d:
             print(f"2D selection ({variant}) {f1:.2%} flag, {f2:.2%} size, "
-                  f"{f3:.2%} SNR")
+                  f"{f3:.2%} SNR, {f4:.2%} any z bin")
+            print("total 2D", sel.sum())
         elif verbose:
```

Comment:
I was little confused, we have an argument on this function `is_2d = True` The only place this is used is here. It seems like we will be applying all these corrections no matter what, so that there is an option confuses me, and it confuses me that there are two different verbose options.

Reply:
Ah, I see what you mean - this is just becuase the select() method calls the select_2d() method to avoid duplication.  select_2d first looks to see if the object is valid for any bin (SNR, size), and select() checks to see if it's in the current z bin.  Without that check you'd get some confusing output.  Does that make sense?

Reply:
what confuses me is we have 2 versions of the verbose statement. One is `verbose` and `is_2d`, this one shows all 4 terms, the other one is just `verbose`: and shows 3 terms. But the code adds all 4 terms to the selection no matter what `is_2d` is

Reply:
Right - is_2d just modifies the verbosity, it records whether the `select_2d` method is being called from within the `select` method or not.  I'll rename to make this a bit clearer!